File: mobile/ios/CallManager.swift
import Foundation
import AVFoundation
import Network
import Combine

// Call Manager - координирует все компоненты звонка
class CallManager: ObservableObject {
    static let shared = CallManager()

    // Published Properties
    @Published var connectionStatus: ConnectionStatus = .disconnected
    @Published var chatMessages: [ChatMessage] = []
    @Published var callStats: CallStatistics = CallStatistics()

    // Components
    var audioCaptureManager: AudioCaptureManager?
    var videoCaptureManager: VideoCaptureManager?
    var networkManager: NetworkManager?
    var codecManager: CodecManager?

    // Private Properties
    private var callStartTime: Date?
    private var statsTimer: Timer?
    private var cancellables = Set<AnyCancelable>()

    init() {
        setupComponents()
    }

    // MARK: - Setup

    func setupComponents() {
        audioCaptureManager = AudioCaptureManager()
        videoCaptureManager = VideoCaptureManager()
        networkManager = NetworkManager()
        codecManager = CodecManager()

        // Setup network callbacks
        networkManager?.onDataReceived = { [weak self] data in
            self?.handleReceivedData(data)
        }

        networkManager?.onConnectionStateChanged = { [weak self] state in
            DispatchQueue.main.async {
                self?.connectionStatus = state
            }
        }
    }

    // MARK: - Connection

    func connect(to address: String) {
        print("Connecting to: \(address)")

        connectionStatus = .connecting

        // Parse address (format: "hostname:port")
        let components = address.split(separator: ":")
        guard components.count == 2,
              let port = UInt16(components[1]) else {
            print("Invalid address format")
            connectionStatus = .disconnected
            return
        }

        let hostname = String(components[0])

        // Start network connection
        networkManager?.connect(host: hostname, port: NWEndpoint.Port(rawValue: port)!) { [weak self] success in
            if success {
                self?.startCall()
            } else {
                DispatchQueue.main.async {
                    self?.connectionStatus = .disconnected
                }
            }
        }
    }

    func disconnect() {
        print("Disconnecting...")

        stopCall()
        networkManager?.disconnect()

        DispatchQueue.main.async {
            self.connectionStatus = .disconnected
            self.callStartTime = nil
            self.chatMessages.removeAll()
        }
    }

    // MARK: - Call Control

    private func startCall() {
        print("Starting call...")

        DispatchQueue.main.async {
            self.connectionStatus = .connected
            self.callStartTime = Date()
        }

        // Start audio capture
        audioCaptureManager?.startCapture { [weak self] audioData in
            self?.handleCapturedAudio(audioData)
        }

        // Start video capture
        videoCaptureManager?.startCapture { [weak self] videoFrame in
            self?.handleCapturedVideo(videoFrame)
        }

        // Start statistics timer
        startStatsTimer()
    }

    private func stopCall() {
        print("Stopping call...")

        audioCaptureManager?.stopCapture()
        videoCaptureManager?.stopCapture()
        statsTimer?.invalidate()
        statsTimer = nil
    }

    // MARK: - Audio/Video Handling

    private func handleCapturedAudio(_ audioData: Data) {
        // Encode audio
        guard let encodedData = codecManager?.encodeAudio(audioData) else {
            return
        }

        // Send over network
        networkManager?.send(encodedData, type: .audio)
    }

    private func handleCapturedVideo(_ videoFrame: CVPixelBuffer) {
        // Convert to data
        guard let imageData = videoCaptureManager?.convertPixelBufferToData(videoFrame) else {
            return
        }

        // Encode video
        guard let encodedData = codecManager?.encodeVideo(imageData) else {
            return
        }

        // Send over network
        networkManager?.send(encodedData, type: .video)
    }

    private func handleReceivedData(_ data: Data) {
        // Parse packet header to determine type
        guard data.count > 4 else { return }

        let packetType = data[0]

        switch packetType {
        case 0x01: // Audio packet
            handleReceivedAudio(data)
        case 0x02: // Video packet
            handleReceivedVideo(data)
        case 0x03: // Chat message
            handleReceivedChatMessage(data)
        default:
            print("Unknown packet type: \(packetType)")
        }
    }

    private func handleReceivedAudio(_ data: Data) {
        // Decode audio
        guard let decodedData = codecManager?.decodeAudio(data) else {
            return
        }

        // Play audio
        audioCaptureManager?.playback(decodedData)
    }

    private func handleReceivedVideo(_ data: Data) {
        // Decode video
        guard let decodedData = codecManager?.decodeVideo(data) else {
            return
        }

        // Display video (placeholder)
        // In full implementation, would update UI with decoded frame
    }

    private func handleReceivedChatMessage(_ data: Data) {
        // Extract message text
        guard let messageText = String(data: data.dropFirst(), encoding: .utf8) else {
            return
        }

        DispatchQueue.main.async {
            let message = ChatMessage(
                text: messageText,
                timestamp: Date(),
                isSent: false
            )
            self.chatMessages.append(message)
        }
    }

    // MARK: - Controls

    func setMuted(_ muted: Bool) {
        audioCaptureManager?.setMuted(muted)
    }

    func setCameraEnabled(_ enabled: Bool) {
        if enabled {
            videoCaptureManager?.resumeCapture()
        } else {
            videoCaptureManager?.pauseCapture()
        }
    }

    func sendChatMessage(_ text: String) {
        // Create chat message
        let message = ChatMessage(
            text: text,
            timestamp: Date(),
            isSent: true
        )

        DispatchQueue.main.async {
            self.chatMessages.append(message)
        }

        // Send over network
        var data = Data([0x03]) // Chat packet type
        if let textData = text.data(using: .utf8) {
            data.append(textData)
        }

        networkManager?.send(data, type: .chat)
    }

    // MARK: - Settings

    func updateSettings(sampleRate: Int, bitrate: Int, enableVAD: Bool, enableNoiseSuppression: Bool) {
        audioCaptureManager?.updateSettings(
            sampleRate: sampleRate,
            enableVAD: enableVAD,
            enableNoiseSuppression: enableNoiseSuppression
        )

        codecManager?.updateSettings(bitrate: bitrate)
    }

    // MARK: - Statistics

    private func startStatsTimer() {
        statsTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.updateStatistics()
        }
    }

    private func updateStatistics() {
        guard let startTime = callStartTime else { return }

        DispatchQueue.main.async {
            self.callStats.duration = Date().timeIntervalSince(startTime)

            // Get network stats
            if let networkStats = self.networkManager?.getStatistics() {
                self.callStats.bitrate = networkStats.bitrate
                self.callStats.packetLoss = networkStats.packetLoss
                self.callStats.rtt = networkStats.rtt
            }
        }
    }
}

// MARK: - Audio Capture Manager

class AudioCaptureManager {
    private var audioEngine: AVAudioEngine?
    private var inputNode: AVAudioInputNode?
    private var isMuted = false
    private var sampleRate: Double = 16000
    private var onAudioCaptured: ((Data) -> Void)?

    func startCapture(onAudioCaptured: @escaping (Data) -> Void) {
        self.onAudioCaptured = onAudioCaptured

        audioEngine = AVAudioEngine()
        guard let audioEngine = audioEngine else { return }

        inputNode = audioEngine.inputNode

        // Configure audio format
        let format = AVAudioFormat(
            commonFormat: .pcmFormatInt16,
            sampleRate: sampleRate,
            channels: 1,
            interleaved: false
        )!

        // Install tap
        inputNode?.installTap(onBus: 0, bufferSize: 320, format: format) { [weak self] buffer, time in
            guard let self = self, !self.isMuted else { return }
            self.processAudioBuffer(buffer)
        }

        // Start engine
        do {
            try audioEngine.start()
            print("Audio capture started")
        } catch {
            print("Failed to start audio engine: \(error)")
        }
    }

    func stopCapture() {
        inputNode?.removeTap(onBus: 0)
        audioEngine?.stop()
        audioEngine = nil
        inputNode = nil
    }

    func setMuted(_ muted: Bool) {
        isMuted = muted
    }

    func updateSettings(sampleRate: Int, enableVAD: Bool, enableNoiseSuppression: Bool) {
        self.sampleRate = Double(sampleRate)
        // Would restart capture with new settings
    }

    func playback(_ audioData: Data) {
        // Playback implementation (simplified)
        // In full implementation, would use AVAudioPlayerNode
    }

    private func processAudioBuffer(_ buffer: AVAudioPCMBuffer) {
        guard let channelData = buffer.int16ChannelData else { return }

        let frameLength = Int(buffer.frameLength)
        let data = Data(bytes: channelData[0], count: frameLength * MemoryLayout<Int16>.size)

        onAudioCaptured?(data)
    }
}

// MARK: - Video Capture Manager

class VideoCaptureManager: NSObject, AVCaptureVideoDataOutputSampleBufferDelegate {
    private var captureSession: AVCaptureSession?
    private var previewLayer: AVCaptureVideoPreviewLayer?
    private var videoOutput: AVCaptureVideoDataOutput?
    private var onVideoCaptured: ((CVPixelBuffer) -> Void)?
    private var isPaused = false

    func setupPreview(in view: UIView) {
        guard let captureSession = captureSession else { return }

        previewLayer = AVCaptureVideoPreviewLayer(session: captureSession)
        previewLayer?.videoGravity = .resizeAspectFill
        previewLayer?.frame = view.bounds

        if let previewLayer = previewLayer {
            view.layer.addSublayer(previewLayer)
        }
    }

    func startCapture(onVideoCaptured: @escaping (CVPixelBuffer) -> Void) {
        self.onVideoCaptured = onVideoCaptured

        captureSession = AVCaptureSession()
        guard let captureSession = captureSession else { return }

        captureSession.beginConfiguration()

        // Add video input
        guard let videoDevice = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .front),
              let videoInput = try? AVCaptureDeviceInput(device: videoDevice),
              captureSession.canAddInput(videoInput) else {
            print("Failed to add video input")
            return
        }

        captureSession.addInput(videoInput)

        // Add video output
        videoOutput = AVCaptureVideoDataOutput()
        videoOutput?.setSampleBufferDelegate(self, queue: DispatchQueue(label: "videoQueue"))

        if let videoOutput = videoOutput,
           captureSession.canAddOutput(videoOutput) {
            captureSession.addOutput(videoOutput)
        }

        captureSession.commitConfiguration()

        // Start session
        DispatchQueue.global(qos: .userInitiated).async {
            captureSession.startRunning()
        }

        print("Video capture started")
    }

    func stopCapture() {
        captureSession?.stopRunning()
        captureSession = nil
    }

    func pauseCapture() {
        isPaused = true
    }

    func resumeCapture() {
        isPaused = false
    }

    func convertPixelBufferToData(_ pixelBuffer: CVPixelBuffer) -> Data? {
        // Convert CVPixelBuffer to Data (simplified)
        // Full implementation would use proper image conversion
        return Data()
    }

    // AVCaptureVideoDataOutputSampleBufferDelegate
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        guard !isPaused,
              let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            return
        }

        onVideoCaptured?(pixelBuffer)
    }
}

// MARK: - Network Manager

class NetworkManager {
    private var connection: NWConnection?
    private var receiveQueue = DispatchQueue(label: "receiveQueue")

    var onDataReceived: ((Data) -> Void)?
    var onConnectionStateChanged: ((ConnectionStatus) -> Void)?

    func connect(host: String, port: NWEndpoint.Port, completion: @escaping (Bool) -> Void) {
        let endpoint = NWEndpoint.hostPort(host: NWEndpoint.Host(host), port: port)
        connection = NWConnection(to: endpoint, using: .udp)

        connection?.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready:
                print("Network connection ready")
                self?.onConnectionStateChanged?(.connected)
                self?.startReceiving()
                completion(true)
            case .failed(let error):
                print("Network connection failed: \(error)")
                self?.onConnectionStateChanged?(.disconnected)
                completion(false)
            default:
                break
            }
        }

        connection?.start(queue: receiveQueue)
    }

    func disconnect() {
        connection?.cancel()
        connection = nil
    }

    func send(_ data: Data, type: PacketType) {
        connection?.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                print("Send error: \(error)")
            }
        })
    }

    private func startReceiving() {
        connection?.receiveMessage { [weak self] data, context, isComplete, error in
            if let data = data {
                self?.onDataReceived?(data)
            }

            // Continue receiving
            self?.startReceiving()
        }
    }

    func getStatistics() -> NetworkStatistics {
        // Return mock statistics
        return NetworkStatistics(bitrate: 128, packetLoss: 0.5, rtt: 50)
    }
}

// MARK: - Codec Manager

class CodecManager {
    private var audioBitrate = 128000
    private var videoBitrate = 1000000

    func encodeAudio(_ data: Data) -> Data? {
        // Simplified audio encoding (Opus would be used in production)
        return data
    }

    func decodeAudio(_ data: Data) -> Data? {
        // Simplified audio decoding
        return data.dropFirst() // Remove packet type byte
    }

    func encodeVideo(_ data: Data) -> Data? {
        // Simplified video encoding (H.264 would be used in production)
        return data
    }

    func decodeVideo(_ data: Data) -> Data? {
        // Simplified video decoding
        return data.dropFirst() // Remove packet type byte
    }

    func updateSettings(bitrate: Int) {
        self.audioBitrate = bitrate * 1000
    }
}

// MARK: - Supporting Types

enum PacketType {
    case audio
    case video
    case chat
}

struct NetworkStatistics {
    let bitrate: Int
    let packetLoss: Double
    let rtt: Int
}

File: mobile/android/CallManager.kt
package com.tvcp.mobile

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.media.*
import android.view.SurfaceView
import android.view.View
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.*
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.nio.ByteBuffer
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.concurrent.timer

// Call Manager
class CallManager(private val context: Context) {
    private val audioCaptureManager = AudioCaptureManager()
    private val videoCaptureManager = VideoCaptureManager(context)
    private val networkManager = NetworkManager()
    private val codecManager = CodecManager()

    private var callStartTime: Long = 0
    private var statsTimer: Timer? = null

    // Callbacks
    var onConnectionStateChanged: ((ConnectionStatus) -> Unit)? = null
    var onStatsUpdated: ((CallStatistics) -> Unit)? = null
    var onChatMessageReceived: ((ChatMessage) -> Unit)? = null

    init {
        setupNetworkCallbacks()
    }

    private fun setupNetworkCallbacks() {
        networkManager.onDataReceived = { data ->
            handleReceivedData(data)
        }

        networkManager.onConnectionStateChanged = { state ->
            onConnectionStateChanged?.invoke(state)
        }
    }

    fun connect(address: String) {
        val parts = address.split(":")
        if (parts.size != 2) {
            println("Invalid address format")
            return
        }

        val host = parts[0]
        val port = parts[1].toIntOrNull() ?: return

        networkManager.connect(host, port) { success ->
            if (success) {
                startCall()
            }
        }
    }

    fun disconnect() {
        stopCall()
        networkManager.disconnect()
        onConnectionStateChanged?.invoke(ConnectionStatus.DISCONNECTED)
    }

    private fun startCall() {
        callStartTime = System.currentTimeMillis()

        // Start audio capture
        audioCaptureManager.startCapture { audioData ->
            handleCapturedAudio(audioData)
        }

        // Start video capture
        videoCaptureManager.startCapture { videoFrame ->
            handleCapturedVideo(videoFrame)
        }

        // Start statistics timer
        startStatsTimer()

        onConnectionStateChanged?.invoke(ConnectionStatus.CONNECTED)
    }

    private fun stopCall() {
        audioCaptureManager.stopCapture()
        videoCaptureManager.stopCapture()
        statsTimer?.cancel()
        statsTimer = null
    }

    private fun handleCapturedAudio(audioData: ByteArray) {
        // Encode audio
        val encodedData = codecManager.encodeAudio(audioData)

        // Send over network
        networkManager.send(encodedData, PacketType.AUDIO)
    }

    private fun handleCapturedVideo(videoFrame: ByteArray) {
        // Encode video
        val encodedData = codecManager.encodeVideo(videoFrame)

        // Send over network
        networkManager.send(encodedData, PacketType.VIDEO)
    }

    private fun handleReceivedData(data: ByteArray) {
        if (data.isEmpty()) return

        when (data[0].toInt()) {
            0x01 -> handleReceivedAudio(data)
            0x02 -> handleReceivedVideo(data)
            0x03 -> handleReceivedChatMessage(data)
        }
    }

    private fun handleReceivedAudio(data: ByteArray) {
        // Decode audio
        val decodedData = codecManager.decodeAudio(data)

        // Play audio
        audioCaptureManager.playback(decodedData)
    }

    private fun handleReceivedVideo(data: ByteArray) {
        // Decode video
        val decodedData = codecManager.decodeVideo(data)

        // Display video (would update UI)
    }

    private fun handleReceivedChatMessage(data: ByteArray) {
        val messageText = String(data.sliceArray(1 until data.size))
        val message = ChatMessage(
            text = messageText,
            timestamp = Date(),
            isSent = false
        )

        onChatMessageReceived?.invoke(message)
    }

    fun setMuted(muted: Boolean) {
        audioCaptureManager.setMuted(muted)
    }

    fun setCameraEnabled(enabled: Boolean) {
        if (enabled) {
            videoCaptureManager.resumeCapture()
        } else {
            videoCaptureManager.pauseCapture()
        }
    }

    fun sendChatMessage(text: String) {
        val data = byteArrayOf(0x03) + text.toByteArray()
        networkManager.send(data, PacketType.CHAT)
    }

    fun updateSettings(settings: CallSettings) {
        audioCaptureManager.updateSettings(
            sampleRate = settings.sampleRate,
            enableVAD = settings.enableVAD,
            enableNoiseSuppression = settings.enableNoiseSuppression
        )

        codecManager.updateSettings(bitrate = settings.bitrate)
    }

    fun getVideoView(context: Context): View {
        return videoCaptureManager.getPreviewView(context)
    }

    private fun startStatsTimer() {
        statsTimer = timer(period = 1000) {
            updateStatistics()
        }
    }

    private fun updateStatistics() {
        val duration = (System.currentTimeMillis() - callStartTime) / 1000
        val networkStats = networkManager.getStatistics()

        val stats = CallStatistics(
            duration = duration,
            bitrate = networkStats.bitrate,
            packetLoss = networkStats.packetLoss,
            rtt = networkStats.rtt
        )

        onStatsUpdated?.invoke(stats)
    }
}

// Audio Capture Manager
class AudioCaptureManager {
    private var audioRecord: AudioRecord? = null
    private var audioTrack: AudioTrack? = null
    private var isCapturing = false
    private var isMuted = false
    private var captureThread: Thread? = null
    private var onAudioCaptured: ((ByteArray) -> Unit)? = null

    private val sampleRate = 16000
    private val channelConfig = AudioFormat.CHANNEL_IN_MONO
    private val audioFormat = AudioFormat.ENCODING_PCM_16BIT
    private val bufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig, audioFormat)

    fun startCapture(onAudioCaptured: (ByteArray) -> Void) {
        this.onAudioCaptured = onAudioCaptured

        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRate,
            channelConfig,
            audioFormat,
            bufferSize
        )

        // Setup audio track for playback
        audioTrack = AudioTrack(
            AudioManager.STREAM_VOICE_CALL,
            sampleRate,
            AudioFormat.CHANNEL_OUT_MONO,
            audioFormat,
            bufferSize,
            AudioTrack.MODE_STREAM
        )

        audioTrack?.play()

        isCapturing = true
        audioRecord?.startRecording()

        captureThread = Thread {
            captureLoop()
        }
        captureThread?.start()
    }

    fun stopCapture() {
        isCapturing = false
        captureThread?.join()

        audioRecord?.stop()
        audioRecord?.release()
        audioRecord = null

        audioTrack?.stop()
        audioTrack?.release()
        audioTrack = null
    }

    fun setMuted(muted: Boolean) {
        isMuted = muted
    }

    fun updateSettings(sampleRate: Int, enableVAD: Boolean, enableNoiseSuppression: Boolean) {
        // Would restart capture with new settings
    }

    fun playback(audioData: ByteArray) {
        audioTrack?.write(audioData, 0, audioData.size)
    }

    private fun captureLoop() {
        val buffer = ByteArray(bufferSize)

        while (isCapturing) {
            val read = audioRecord?.read(buffer, 0, buffer.size) ?: 0

            if (read > 0 && !isMuted) {
                onAudioCaptured?.invoke(buffer.copyOf(read))
            }
        }
    }
}

// Video Capture Manager
class VideoCaptureManager(private val context: Context) {
    private var cameraProvider: ProcessCameraProvider? = null
    private var camera: Camera? = null
    private var previewView: PreviewView? = null
    private var imageAnalyzer: ImageAnalysis? = null
    private var isPaused = false
    private var onVideoCaptured: ((ByteArray) -> Unit)? = null

    private val cameraExecutor: ExecutorService = Executors.newSingleThreadExecutor()

    fun startCapture(onVideoCaptured: (ByteArray) -> Unit) {
        this.onVideoCaptured = onVideoCaptured

        val cameraProviderFuture = ProcessCameraProvider.getInstance(context)

        cameraProviderFuture.addListener({
            cameraProvider = cameraProviderFuture.get()
            bindCamera()
        }, ContextCompat.getMainExecutor(context))
    }

    fun stopCapture() {
        cameraProvider?.unbindAll()
        cameraExecutor.shutdown()
    }

    fun pauseCapture() {
        isPaused = true
    }

    fun resumeCapture() {
        isPaused = false
    }

    fun getPreviewView(context: Context): PreviewView {
        if (previewView == null) {
            previewView = PreviewView(context)
        }
        return previewView!!
    }

    private fun bindCamera() {
        val preview = Preview.Builder().build()

        previewView?.let {
            preview.setSurfaceProvider(it.surfaceProvider)
        }

        // Image analyzer for frame capture
        imageAnalyzer = ImageAnalysis.Builder()
            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            .build()
            .also {
                it.setAnalyzer(cameraExecutor) { imageProxy ->
                    if (!isPaused) {
                        processFrame(imageProxy)
                    }
                    imageProxy.close()
                }
            }

        val cameraSelector = CameraSelector.DEFAULT_FRONT_CAMERA

        try {
            cameraProvider?.unbindAll()
            camera = cameraProvider?.bindToLifecycle(
                context as LifecycleOwner,
                cameraSelector,
                preview,
                imageAnalyzer
            )
        } catch (e: Exception) {
            println("Failed to bind camera: ${e.message}")
        }
    }

    private fun processFrame(imageProxy: ImageProxy) {
        // Convert ImageProxy to ByteArray (simplified)
        val buffer = imageProxy.planes[0].buffer
        val bytes = ByteArray(buffer.remaining())
        buffer.get(bytes)

        onVideoCaptured?.invoke(bytes)
    }
}

// Network Manager
class NetworkManager {
    private var socket: DatagramSocket? = null
    private var remoteAddress: InetAddress? = null
    private var remotePort: Int = 0
    private var isConnected = false
    private var receiveThread: Thread? = null

    var onDataReceived: ((ByteArray) -> Unit)? = null
    var onConnectionStateChanged: ((ConnectionStatus) -> Unit)? = null

    fun connect(host: String, port: Int, callback: (Boolean) -> Unit) {
        try {
            socket = DatagramSocket()
            remoteAddress = InetAddress.getByName(host)
            remotePort = port
            isConnected = true

            // Start receive thread
            receiveThread = Thread {
                receiveLoop()
            }
            receiveThread?.start()

            onConnectionStateChanged?.invoke(ConnectionStatus.CONNECTED)
            callback(true)
        } catch (e: Exception) {
            println("Connection failed: ${e.message}")
            onConnectionStateChanged?.invoke(ConnectionStatus.DISCONNECTED)
            callback(false)
        }
    }

    fun disconnect() {
        isConnected = false
        receiveThread?.join()
        socket?.close()
        socket = null
    }

    fun send(data: ByteArray, type: PacketType) {
        if (!isConnected || socket == null) return

        try {
            val packet = DatagramPacket(
                data,
                data.size,
                remoteAddress,
                remotePort
            )
            socket?.send(packet)
        } catch (e: Exception) {
            println("Send error: ${e.message}")
        }
    }

    private fun receiveLoop() {
        val buffer = ByteArray(65536)

        while (isConnected) {
            try {
                val packet = DatagramPacket(buffer, buffer.size)
                socket?.receive(packet)

                val data = packet.data.copyOf(packet.length)
                onDataReceived?.invoke(data)
            } catch (e: Exception) {
                if (isConnected) {
                    println("Receive error: ${e.message}")
                }
            }
        }
    }

    fun getStatistics(): NetworkStatistics {
        // Return mock statistics
        return NetworkStatistics(
            bitrate = 128,
            packetLoss = 0.5,
            rtt = 50
        )
    }
}

// Codec Manager
class CodecManager {
    private var audioBitrate = 128000
    private var videoBitrate = 1000000

    fun encodeAudio(data: ByteArray): ByteArray {
        // Simplified encoding (Opus would be used in production)
        return byteArrayOf(0x01) + data
    }

    fun decodeAudio(data: ByteArray): ByteArray {
        // Simplified decoding
        return data.sliceArray(1 until data.size)
    }

    fun encodeVideo(data: ByteArray): ByteArray {
        // Simplified encoding (H.264 would be used in production)
        return byteArrayOf(0x02) + data
    }

    fun decodeVideo(data: ByteArray): ByteArray {
        // Simplified decoding
        return data.sliceArray(1 until data.size)
    }

    fun updateSettings(bitrate: Int) {
        this.audioBitrate = bitrate * 1000
    }
}

// ViewModel
class CallViewModel : ViewModel() {
    val connectionStatus = MutableLiveData(ConnectionStatus.DISCONNECTED)
    val callStats = MutableLiveData(CallStatistics())
    val chatMessages = MutableLiveData<List<ChatMessage>>(emptyList())

    fun updateStatus(status: ConnectionStatus) {
        connectionStatus.value = status
    }

    fun updateStats(stats: CallStatistics) {
        callStats.value = stats
    }

    fun addChatMessage(text: String, isSent: Boolean) {
        val message = ChatMessage(
            text = text,
            timestamp = Date(),
            isSent = isSent
        )

        val current = chatMessages.value ?: emptyList()
        chatMessages.value = current + message
    }
}

// Supporting Types
enum class PacketType {
    AUDIO,
    VIDEO,
    CHAT
}

data class NetworkStatistics(
    val bitrate: Int,
    val packetLoss: Double,
    val rtt: Int
)
